### ISSUE TITLE
[internal] The session has been destroyed

### ISSUE THREAD

**Issue by akosyakov** [connectrpc/connect-es#683]
version: 0.10.0

```
ConnectError: [internal] The session has been destroyed
	at i.from (/out/extension.js:2:1820120)
	at t.connectErrorFromNodeReason (/out/extension.js:2:1804959)
	at Promise.reject (/out/extension.js:2:1809452)
	at /out/extension.js:2:1811572
```

Unfortunately stack trace is minified on our side not sure which path lead to that, it seem to be an unary call. It seems somewhere stream is called when it is already destroyed.

**Comment by timostamm** [comment]
The issue might be related to https://github.com/bufbuild/connect-es/pull/681, which was just fixed in [v0.10.1](https://github.com/bufbuild/connect-es/releases/tag/v0.10.1), but it's impossible to tell without more information. 

Can you enable source maps in your build so that it's possible to restore stack traces?

**Comment by akosyakov** [comment]
I tried to recover and got following stack trace:

```
i.from at ../node_modules/@bufbuild/connect/dist/cjs/connect-error.js:73:19
from at ../node_modules/@bufbuild/connect-node/dist/cjs/node-error.js:54:38
reason at ../node_modules/@bufbuild/connect-node/dist/cjs/node-universal-client.js:305:110
reject at ../node_modules/@bufbuild/connect-node/dist/cjs/node-universal-client.js:204:17
```

**Comment by akosyakov** [comment]
so it is coming from here: https://github.com/bufbuild/connect-es/blob/76b17edbbc4c0d4c98d1de8eb754192835458054/packages/connect-node/src/node-universal-client.ts#L340

the rest of stack trace is wrapping an error in connect error

**Comment by timostamm** [comment]
That's good info. Since the stack does not include `gotoReady`, it means the error is raised here:

https://github.com/bufbuild/connect-es/blob/b6187e9b2007bf178c0a6dea79cab3f30ff09c1a/packages/connect-node/src/http2-session-manager.ts#L201

This means the connection owned by the [ready state](https://github.com/bufbuild/connect-es/blob/b6187e9b2007bf178c0a6dea79cab3f30ff09c1a/packages/connect-node/src/http2-session-manager.ts#L510) has been destroyed without the state noticing. Looking at the source, I don't see a smoking gun, we're maintaining event listeners on "close" and "error" all the time.

It's also possible that the connection is already dead if we enter the "ready" state. We noticed some unexpected behavior with the `http2` module while the connection is established, see https://github.com/bufbuild/connect-es/blob/b6187e9b2007bf178c0a6dea79cab3f30ff09c1a/packages/connect-node/src/http2-session-manager.ts#L415-L417
https://github.com/bufbuild/connect-es/pull/687 adds an assertion so that we can identify whether this is the case.

**Comment by akosyakov** [comment]
fyi we upgrade to 0.10.1 but we still can see it happens for some users, so it does not seem to go away with https://github.com/bufbuild/connect-es/pull/681

**Comment by akosyakov** [comment]
@timostamm We also still see `New streams cannot be created after receiving a GOAWAY` as internal errors. Are they expected to be thrown happen after recent fixes? as internal?

**Comment by timostamm** [comment]
@akosyakov, with #681, a new call will open a new connection if the current connection has received a GOAWAY. The error you see is unexpected. Will update #680 with the new information.

**Comment by timostamm** [comment]
[v0.12.0](https://github.com/bufbuild/connect-es/releases/tag/v0.11.1) was just published with https://github.com/bufbuild/connect-es/pull/712.

We can reproduce the fix with the following script that uses the session manager:

<details>
<summary>repro.mjs</summary>

```js
import * as http2 from "http2";
import { createNodeHttpClient, Http2SessionManager } from "@bufbuild/connect-node";
import assert from "assert";

// start a server that sends a GOAWAY with INTERNAL ERROR to every session
const PORT = 8088;
const server = http2.createServer()
    .on("session", (sess) => {
        setTimeout(() => {
            sess.goaway(http2.constants.NGHTTP2_INTERNAL_ERROR)
        }, 100);
    })
    .on("request", (req, res) => {
        res.writeHead(200);
        setTimeout(() => res.end(), 200);
    });
await new Promise(resolve => {
    server.listen(PORT, () => resolve());
});


// create an h2 client
const sm = new Http2SessionManager(`http://localhost:${PORT}`);
const client = createNodeHttpClient({
    httpVersion: "2",
    sessionProvider: () => sm
});


// make requests
const res1 = await client({
    url: `http://localhost:${PORT}`,
    method: "POST",
    header: new Headers(),
});
try {
    for await (const chunk of res1.body) {
        // consume response body to see error
    }
} catch (e) {
    assert(String(e) === "ConnectError: [internal] Session closed with error code 2");
}

console.log("res1 done")

const res2 = await client({
    url: `http://localhost:${PORT}`,
    method: "POST",
    header: new Headers(),
});

console.log("res2 done")
```
</details>

```bash


```

**Comment by jeanp413** [comment]
@timostamm we released a new version of our extension with the connect library updated to `0.0.12` and the errors are back 😢 , could you reopen this 🙏

**Comment by timostamm** [comment]
The new fix just shipped in [v0.13.2](https://github.com/connectrpc/connect-es/releases/tag/v0.13.2). Let us know if this fixes the problem.
